packages/lesswrong/server/fmCrosspost.ts

```diff
@@ -0,0 +1,381 @@
+import fetch from "node-fetch";
+import jwt from "jsonwebtoken";
+import bodyParser from "body-parser";
+import type { Application, Request, Response } from "express";
+
+const crosspostSigningKeySetting = new DatabaseServerSetting<string|null>("fmCrosspostSigningKey", null);
+
+const jwtSigningOptions = {
+  algorithm: "HS256",
+  expiresIn: "30m",
+} as const;
+
+const apiRoutes = {
+  crosspostToken: "/api/crosspostToken",
+  connectCrossposter: "/api/connectCrossposter",
+  unlinkCrossposter: "/api/unlinkCrossposter",
+  crosspost: "/api/crosspost",
+  updateCrosspost: "/api/updateCrosspost",
+} as const;
+
+type ApiRoute = typeof apiRoutes[keyof typeof apiRoutes];
+
+type ConnectCrossposterArgs = {
+  token: string,
+}
+
+type ConnectCrossposterPayload = {
+  userId: string,
+}
+
+type UnlinkCrossposterPayload = {
+  userId: string,
+}
+
+type UpdateCrosspostPayload = {
+  postId: string,
+  draft: boolean,
+  deletedDraft: boolean,
+  title: string,
+}
+
+type CrosspostPayload = {
+  localUserId: string,
+  foreignUserId: string,
+}
+
+type Crosspost = Pick<DbPost, "_id" | "title" | "userId" | "fmCrosspost" | "draft">;
+
+class ApiError extends Error {
+  constructor(public code: number, message: string) {
+    super(message);
+  }
+}
+
+class UnauthorizedError extends ApiError {
+  constructor() {
+    super(403, "You must login to do this");
+  }
+}
+
+class MissingSecretError extends ApiError {
+  constructor() {
+    super(500, "Missing crosspost signing secret env var");
+  }
+}
+
+class MissingParametersError extends ApiError {
+  constructor(expectedParams: string[], body: any) {
+    super(400, `Missing parameters: expected ${JSON.stringify(expectedParams)} but received ${JSON.stringify(body)}`);
+  }
+}
+
+class InvalidTokenError extends ApiError {
+  constructor() {
+    super(400, "Invalid token");
+  }
+}
+
+class InvalidUserError extends ApiError {
+  constructor() {
+    super(400, "Invalid user");
+  }
+}
+
+const getSecret = () => {
+  const secret = crosspostSigningKeySetting.get();
+  if (!secret) {
+    throw new MissingSecretError();
+  }
+  return secret;
+}
+
+const getUserId = (req?: Request) => {
+  const userId = req?.user?._id;
+  if (!userId) {
+    throw new UnauthorizedError();
+  }
+  return userId;
+}
+
+const makeApiUrl = (route: ApiRoute) => fmCrosspostBaseUrlSetting.get() + route.slice(1);
+
+const makeCrossSiteRequest = async <T extends {}>(
+  route: ApiRoute,
+  body: T,
+  expectedStatus: string,
+  onErrorMessage: string,
+) => {
+  const result = await fetch(makeApiUrl(route), {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify(body),
+  });
+  const json = await result.json();
+  if (json.status !== expectedStatus) {
+    throw new ApiError(500, onErrorMessage);
+  }
+  return json;
+}
+
+const signToken = <T extends {}>(payload: T): string =>
+  jwt.sign(payload, getSecret(), jwtSigningOptions);
```

Comment:
I think this is the synchronous version of the function - we should probably do this async, since we don't want to block the thread on signing/verifying (if nothing else, it'd make it much easier to DOS us).  It looks like the `jsonwebtoken` library only provides an async version using callbacks, but we can just `promisify` it.

Reply:
Agreed. I ended up just wrapping it in a Promise manually instead of using promisify though because the library functions take a variable number of arguments and using promisify made the type checker pretty unhappy. I would've been possible to fix it, but I think it would've been much uglier.